nexus/saga-recovery/src/recovery.rs

```diff
@@ -0,0 +1,705 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Guts of the saga recovery bookkeeping
+
+use super::status::RecoveryFailure;
+use super::status::RecoverySuccess;
+use chrono::{DateTime, Utc};
+use omicron_common::api::external::Error;
+use slog::{debug, error, info, warn};
+use slog_error_chain::InlineErrorChain;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use steno::SagaId;
+use tokio::sync::mpsc;
+
+/// Describes state related to saga recovery that needs to be maintained across
+/// multiple passes
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RestState {
+    /// set of sagas that we believe may be running
+    ///
+    /// See the big block comment in the saga recovery background task for more
+    /// on how this works and why.
+    sagas_started: BTreeMap<SagaId, SagaStartInfo>,
+    remove_next: BTreeSet<SagaId>,
```

Comment:
i notice we use `BTreeMap`s and sets throughout this module, rather than `HashMap`s/`HashSet`s. out of curiosity, is there a particular reason we use ordered collections here? i.e., it would be nice to know if there are any invariants that actually _rely_ on these having consistent orders --- AFAICT, there doesn't seem to be, which is why i'm asking.

Reply:
I don't think the ordering matters in most of these cases, though it does make testing simpler.

I generally reach for the ordered containers by default.  (I've also had to convert various uses of `HashMap` for cases where the order turned out to matter...)